File: app/services/game-analysis.service.ts
import { Observable } from '@nativescript/core';
import { GameState } from '../models/game-state.model';
import { HeroService } from './hero.service';
import { BuildService } from './build.service';

export class GameAnalysisService extends Observable {
  private heroService: HeroService;
  private buildService: BuildService;
  private currentGameState: GameState | null = null;

  constructor() {
    super();
    this.heroService = new HeroService();
    this.buildService = new BuildService();
  }

  analyzeGameState(gameState: GameState): {
    recommendations: string[];
    threats: string[];
    objectives: string[];
  } {
    const recommendations: string[] = [];
    const threats: string[] = [];
    const objectives: string[] = [];

    // Analyze gold difference
    const goldDiff = gameState.allyTeam.totalGold - gameState.enemyTeam.totalGold;
    if (goldDiff < -2000) {
      recommendations.push('Farm more and avoid unnecessary fights');
    } else if (goldDiff > 2000) {
      recommendations.push('Press your advantage and secure objectives');
    }

    // Analyze enemy threats
    gameState.enemyTeam.players.forEach(player => {
      const hero = this.heroService.getHeroById(player.heroId);
      if (hero) {
        if (player.killCount > 3) {
          threats.push(`${hero.name} is fed (${player.killCount} kills)`);
        }
      }
    });

    // Analyze objectives
    if (gameState.gameTime > 420 && gameState.currentObjective === 'lord') {
      objectives.push('Lord is spawning soon, prepare vision');
    }

    return { recommendations, threats, objectives };
  }

  getCounterBuildSuggestions(
    heroId: string,
    enemyTeam: string[]
  ): { itemSuggestions: string[]; reasoning: string[] } {
    const itemSuggestions: string[] = [];
    const reasoning: string[] = [];

    // Analyze enemy team composition
    const hasPhysicalDamage = enemyTeam.some(id => {
      const hero = this.heroService.getHeroById(id);
      return hero?.role === 'Marksman' || hero?.role === 'Fighter';
    });

    const hasMagicDamage = enemyTeam.some(id => {
      const hero = this.heroService.getHeroById(id);
      return hero?.role === 'Mage';
    });

    if (hasPhysicalDamage) {
      itemSuggestions.push('Antique Cuirass');
      reasoning.push('Enemy has physical damage dealers');
    }

    if (hasMagicDamage) {
      itemSuggestions.push('Athena\'s Shield');
      reasoning.push('Enemy has magic damage dealers');
    }

    return { itemSuggestions, reasoning };
  }

  updateGameState(newState: GameState): void {
    this.currentGameState = newState;
    this.notifyPropertyChange('gameState', newState);
  }

  getCurrentGameState(): GameState | null {
    return this.currentGameState;
  }
}

File: src/main/kotlin/com/mobaassistant/service/GameAnalysisService.kt
package com.mobaassistant.service

import com.mobaassistant.model.GameState
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import org.springframework.stereotype.Service
import kotlin.math.abs

@Service
class GameAnalysisService(
    private val heroService: HeroService,
    private val combatAnalysisService: CombatAnalysisService
) {
    fun analyzeGameState(gameState: GameState): GameAnalysis {
        val recommendations = mutableListOf<String>()
        val threats = mutableListOf<String>()
        val objectives = mutableListOf<String>()

        // Analyze gold difference
        val goldDiff = gameState.allyTeam.totalGold - gameState.enemyTeam.totalGold
        when {
            goldDiff < -2000 -> recommendations.add("Farm more and avoid unnecessary fights")
            goldDiff > 2000 -> recommendations.add("Press your advantage and secure objectives")
        }

        // Analyze enemy threats
        gameState.enemyTeam.players.forEach { player ->
            if (player.killCount > 3) {
                threats.add("Enemy ${player.heroId} is fed (${player.killCount} kills)")
            }
        }

        // Analyze objectives
        if (gameState.gameTime > 420 && gameState.currentObjective == "lord") {
            objectives.add("Lord is spawning soon, prepare vision")
        }

        return GameAnalysis(recommendations, threats, objectives)
    }

    fun getCounterBuildSuggestions(
        heroId: String,
        enemyTeam: List<String>
    ): BuildSuggestions {
        val itemSuggestions = mutableListOf<String>()
        val reasoning = mutableListOf<String>()

        // Analyze enemy team composition
        val hasPhysicalDamage = enemyTeam.any { enemyId ->
            val hero = heroService.getHeroById(enemyId.toLong())
            hero.role in listOf("Marksman", "Fighter")
        }

        val hasMagicDamage = enemyTeam.any { enemyId ->
            val hero = heroService.getHeroById(enemyId.toLong())
            hero.role == "Mage"
        }

        if (hasPhysicalDamage) {
            itemSuggestions.add("Antique Cuirass")
            reasoning.add("Enemy has physical damage dealers")
        }

        if (hasMagicDamage) {
            itemSuggestions.add("Athena's Shield")
            reasoning.add("Enemy has magic damage dealers")
        }

        return BuildSuggestions(itemSuggestions, reasoning)
    }

    fun observeGameState(): Flow<GameState> = flow {
        // In a real implementation, this would connect to the game's API
        // For now, we'll emit mock data
        while (true) {
            // Emit game state updates
            kotlinx.coroutines.delay(1000)
        }
    }
}

data class GameAnalysis(
    val recommendations: List<String>,
    val threats: List<String>,
    val objectives: List<String>
)

data class BuildSuggestions(
    val itemSuggestions: List<String>,
    val reasoning: List<String>
)